### ISSUE TITLE
[Bug] Workflow.WaitConditionAsync may be blocked in Temporal 1.1.1 and will never be passed even when condition is true

### ISSUE THREAD

**Issue by eukladchikov-bigw** [temporalio/sdk-dotnet#263]
### What are you really trying to do?

I am upgrading our microservice from Temporal 1.0.0 to 1.1.1 and see that some unit tests fail because they stuck on Workflow.WaitConditionAsync, but downgrading back to 1.0.0 or to 1.1.0 solves the issue

### Describe the bug

Workflow.WaitConditionAsync sometimes hangs probably by a blocked thread caused by this pull request https://github.com/temporalio/sdk-dotnet/pull/242 

I created a generic test that reproduces the bug for Temporal 1.1.1 and works fine downgrading to 1.1.0

In my test it is getting stuck on workflow line:
`await Workflow.WaitConditionAsync(() => _preorderCount == n);`

### Minimal Reproduction

I use Visual Studio 2022 to run it
.NET Project file:

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net7.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>

    <IsPackable>false</IsPackable>
  </PropertyGroup>

  <ItemGroup>
	<PackageReference Include="FluentValidation" Version="11.7.1" />
	<PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="7.0.0" />
	<PackageReference Include="Shouldly" Version="4.2.1" />
	<PackageReference Include="Temporalio" Version="1.1.1" />
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.3.2" />
    <PackageReference Include="xunit" Version="2.4.2" />
    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.5">
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
      <PrivateAssets>all</PrivateAssets>
    </PackageReference>
    <PackageReference Include="coverlet.collector" Version="3.1.2">
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
      <PrivateAssets>all</PrivateAssets>
    </PackageReference>
  </ItemGroup>

</Project>
```

Unit test .cs file:

```
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;
using Shouldly;
using System.Collections.Concurrent;
using System.Diagnostics;
using System.Reflection;
using Temporalio.Activities;
using Temporalio.Client;
using Temporalio.Testing;
using Temporalio.Worker;
using Temporalio.Workflows;

namespace TemporalWaitConditionAsyncIssue;

public class UnitTest1
{
	private ServiceProvider? _serviceProvider;
	private TemporalWorker? _temporalWorker;

	public WorkflowEnvironment? WorkflowEnvironment { get; private set; }

	[Fact]
	public async Task Should_WaitForPreorderCount()
	{
		ClearTemporalCache();
		await StartWorkflowEnvironment(false);
		var wfId = Guid.NewGuid().ToString();
		await StartTemporal(wfId);
		var workflowClient = CreateWorkflowClient(wfId);

		// Act
		for (int i = 0; i < GenericWorkflow1.ConditionCount; i++)
		{
			await workflowClient.SignalPreorderConsignmentReady(wfId);
			await Task.Delay(TimeSpan.FromSeconds(1));
		}

		await Task.Delay(TimeSpan.FromSeconds(5));

		// Assert
		GenericWorkflow1.Finished.ShouldBeTrue();
	}

	private void ClearTemporalCache()
	{
		var cachedDefinitionsField = typeof(ActivityDefinition)
			.GetField("CachedDefinitions", BindingFlags.NonPublic | BindingFlags.Static)
			?.GetValue(null);

		var activityCache = cachedDefinitionsField as ConcurrentDictionary<MethodInfo, ActivityDefinition>;
		activityCache?.Clear();
	}
	private async Task StartWorkflowEnvironment(bool withTimeSkipping)
	{
		WorkflowEnvironment = withTimeSkipping
			? await WorkflowEnvironment.StartTimeSkippingAsync(new WorkflowEnvironmentStartTimeSkippingOptions { })
			: await WorkflowEnvironment.StartLocalAsync();
	}

	private async Task StartTemporal(string taskQueue)
	{
		await StartWorkerService(taskQueue, configureOptions: options =>
		{
			options
				.AddWorkflow<GenericWorkflow1>();
		});

		var workflowId = GenericWorkflow1.GetWorkflowId(taskQueue);
		var options = new WorkflowOptions
		{
			Id = workflowId,
			TaskQueue = taskQueue,
		};

		await WorkflowEnvironment!.Client.StartWorkflowAsync<GenericWorkflow1>(
			wf => wf.RunAsync(),
			options);
	}

	private async Task StartWorkerService(
		string taskQueue = "default",
		Action<TemporalWorkerOptions>? configureOptions = null)
	{
		_temporalWorker?.Dispose();

		if (_serviceProvider != null) await _serviceProvider.DisposeAsync();
		if (WorkflowEnvironment == null)
		{
			throw new InvalidOperationException("The workflow environment has not been started. Precede this call with a call to StartWorkflowEnvironment.");
		}

		var serviceCollection = new ServiceCollection();

		if (configureOptions == null)
		{
			_serviceProvider = serviceCollection.BuildServiceProvider();
		}
		else
		{
			_serviceProvider = serviceCollection.BuildServiceProvider();

			var workerOptions = new TemporalWorkerOptions(taskQueue)
			{
				LoggerFactory = new NullLoggerFactory(),
				DebugMode = Debugger.IsAttached
			};

			configureOptions(workerOptions);

			_temporalWorker = new TemporalWorker(
				WorkflowEnvironment.Client,
				workerOptions);

			// Do not await this task as it will block the test execution
#pragma warning disable CS4014
			_temporalWorker.ExecuteAsync(CancellationToken.None);
#pragma warning restore CS4014
		}
	}

	private GenericWorkflowClient CreateWorkflowClient(string taskQueue = "default")
	{
		if (WorkflowEnvironment == null)
		{
			throw new InvalidOperationException("The workflow environment has not been started. Precede this call with a call to StartWorkflowEnvironment.");
		}

		return new GenericWorkflowClient(WorkflowEnvironment.Client);

	}
}

[Workflow]
public class GenericWorkflow1
{
	public const int ConditionCount = 2;
	public static bool Finished = false;

	private int _preorderCount = 0;

	public static string GetWorkflowId(string orderId)
	{
		return $"{orderId}";
	}

	[WorkflowRun]
	public async Task RunAsync()
	{
		int n = ConditionCount;

		for (int i = 0; i < n; i++)
		{
			await Workflow.WaitConditionAsync(() => _preorderCount > i && _preorderCount <= n);
		}

		await Workflow.WaitConditionAsync(() => _preorderCount == n);

		Finished = true;

	}

	[WorkflowSignal]
	public Task PreorderReady()
	{
		_preorderCount++;
		return Task.CompletedTask;
	}
}

public class GenericWorkflowClient
{
	private readonly ITemporalClient _client;

	public GenericWorkflowClient(ITemporalClient client)
	{
		_client = client;
	}

	public async Task SignalPreorderConsignmentReady(string wfId)
	{
		var workflowId = GenericWorkflow1.GetWorkflowId(wfId);

		await _client.StartWorkflowAsync<GenericWorkflow1>(
				wf => wf.RunAsync(),
				new WorkflowOptions(workflowId, wfId)
				{
					StartSignal = nameof(GenericWorkflow1.PreorderReady),
					StartSignalArgs = Array.Empty<object>()
				});
	}
}
```

### Environment/Versions

- OS and processor: x86 Windows
- Temporalio Version 1.1.1
- Using nuget package and Visual Studio 2022

### Additional context

Suspect these changes https://github.com/temporalio/sdk-dotnet/pull/242

**Comment by cretz** [comment]
Yes, this is a known bug we accidentally caused. You can downgrade to 1.1.0 or wait until an patch release (today or tomorrow) coming when #259 is merged.

**Comment by cretz** [comment]
This should now be fixed with 1.1.2.
